model/historical_test_data.go

```diff
@@ -258,3 +269,484 @@ func (i *HistoricalTestDataInfo) validate() error {
+func buildAddFieldsDateStage(fieldName string, inputDateFieldName string, start time.Time, end time.Time, numDays int) bson.M {
+	inputDateFieldRef := "$" + inputDateFieldName
+	if numDays <= 1 {
+		return bson.M{"$addFields": bson.M{fieldName: inputDateFieldRef}}
+	}
+	boundaries := dateBoundaries(start, end, numDays)
+	branches := make([]bson.M, len(boundaries))
+	for i := 0; i < len(boundaries)-1; i++ {
+		branches[i] = bson.M{
+			"case": bson.M{
+				"$and": []interface{}{
+					bson.M{"$gte": []interface{}{inputDateFieldRef, boundaries[i]}},
+					bson.M{"$lt": []interface{}{inputDateFieldRef, boundaries[i+1]}},
+				},
+			},
+			"then": boundaries[i],
+		}
+	}
+	lastIndex := len(boundaries) - 1
+	branches[lastIndex] = bson.M{
+		"case": bson.M{"$gte": []interface{}{inputDateFieldRef, boundaries[lastIndex]}},
+		"then": boundaries[lastIndex],
+	}
+	return bson.M{"$addFields": bson.M{fieldName: bson.M{"$switch": bson.M{"branches": branches}}}}
+}
+
+// buildHTDGroupId builds the _id field for the $group stage corresponding to
+// the HTDGroupBy value.
+func buildHTDGroupId(groupBy HTDGroupBy) bson.M {
+	id := bson.M{aggregatedHistoricalTestDataDateKey: "$" + historicalTestDataInfoDateKey}
+	switch groupBy {
+	case HTDGroupByVariant:
+		id[aggregatedHistoricalTestDataVariantKey] = "$" + bsonutil.GetDottedKeyName(historicalTestDataInfoKey, historicalTestDataInfoVariantKey)
+		fallthrough
+	case HTDGroupByTask:
+		id[aggregatedHistoricalTestDataTaskNameKey] = "$" + bsonutil.GetDottedKeyName(historicalTestDataInfoKey, historicalTestDataInfoTaskNameKey)
+		fallthrough
+	case HTDGroupByTest:
+		id[aggregatedHistoricalTestDataTestNameKey] = "$" + bsonutil.GetDottedKeyName(historicalTestDataInfoKey, historicalTestDataInfoTestNameKey)
+	}
+	return id
+}
+
+// getNextDate returns the date of the grouping period following the one
+// specified in startAt.
+func (f HistoricalTestDataFilter) getNextDate() time.Time {
+	numDays := time.Duration(f.GroupNumDays) * 24 * time.Hour
+	if f.Sort == HTDSortLatestFirst {
+		return f.StartAt.Date.Add(-numDays)
+	} else {
+		return f.StartAt.Date.Add(numDays)
+	}
+}
+
+// buildTestPaginationOrBranches builds an expression for the conditions
+// imposed by the filter StartAt field.
+func (f HistoricalTestDataFilter) buildTestPaginationOrBranches() []bson.M {
+	var dateDescending = f.Sort == HTDSortLatestFirst
+	var nextDate interface{}
+
+	if f.GroupNumDays > 1 {
+		nextDate = f.getNextDate()
+	}
+
+	fields := []htdPaginationField{
+		{
+			field:      bsonutil.GetDottedKeyName(historicalTestDataInfoKey, historicalTestDataInfoDateKey),
+			descending: dateDescending,
+			strict:     true,
+			value:      f.StartAt.Date,
+			nextValue:  nextDate,
+		},
+	}
+	switch f.GroupBy {
+	case HTDGroupByVariant:
+		fields = append(fields, htdPaginationField{
+			field:  bsonutil.GetDottedKeyName(historicalTestDataInfoKey, historicalTestDataInfoVariantKey),
+			strict: true,
+			value:  f.StartAt.Variant,
+		})
+		fallthrough
```

Comment:
Just to check - is the assumption that, if you group by variant, you also have to specify a task and a test? And similarly, if you group by task, you also have to specify a test?

Reply:
You always have to specify a task and test (see the validation functions above). But if you group by variant, it will also group by task and test, if you group by task it also groups by test, and if you group by test it just groups by test. That is always why variant and task can be returned emtpy.